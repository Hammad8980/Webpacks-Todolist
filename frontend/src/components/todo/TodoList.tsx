import { useTodos } from '../../features/todos/hooks/useTodos';
import { type Task } from '../../features/todos/TodoTaskTypes';
import TodoItem from './TodoItem';

type TodoListProps = {
  tasks: Task[];
  onDelete: (id: number) => void;
  onToggle: (id: number) => void;
  onUpdate: (id: number, title: string) => void;
};

export default function TodoList({
  tasks,
  onDelete,
  onToggle,
  onUpdate,
}: TodoListProps) {
  const {
    isLoading,
  } = useTodos();

  if (tasks.length === 0 && !isLoading) {
    return (
      <div className="text-center py-12">
        <div className="mb-4">
          <svg
            className="w-16 h-16 text-gray-600 mx-auto"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={1.5}
              d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
            />
          </svg>
        </div>
        <p className="text-gray-400 text-lg">No tasks yet!</p>
        <p className="text-gray-500 text-sm mt-1">
          Add your first task above to get started.
        </p>
      </div>
    );
  }

  return (
    <div className="space-y-3">
      {tasks.map((task) => (
        <TodoItem
          key={task.id}
          task={task}
          onDelete={onDelete}
          onToggle={onToggle}
          onUpdate={onUpdate}
        />
      ))}
    </div>
  );
}
