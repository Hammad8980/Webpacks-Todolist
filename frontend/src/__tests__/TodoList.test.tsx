import { render, screen } from '@testing-library/react';
import TodoList from '../components/todo/TodoList';
import type { Task } from '../features/todos/TodoTaskTypes';

// Mock the useTodos hook
jest.mock('../features/todos/hooks/useTodos', () => ({
  useTodos: jest.fn(() => ({
    isLoading: false,
  })),
}));

// Mock TodoItem component to simplify testing
jest.mock('../components/todo/TodoItem', () => {
  return function MockTodoItem({ task }: { task: Task }) {
    return <div data-testid={`todo-item-${task.id}`}>{task.title}</div>;
  };
});

describe('TodoList', () => {
  const mockOnDelete = jest.fn();
  const mockOnToggle = jest.fn();
  const mockOnUpdate = jest.fn();

  const defaultProps = {
    tasks: [] as Task[],
    onDelete: mockOnDelete,
    onToggle: mockOnToggle,
    onUpdate: mockOnUpdate,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('shows empty state when no tasks are provided', () => {
    const { container } = render(<TodoList {...defaultProps} />);
    const svg = document.querySelector('svg');

    expect(screen.getByText('No tasks yet!')).toBeInTheDocument();
    expect(screen.getByText('Add your first task above to get started.')).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    expect(svg).toBeInTheDocument();
  });

  it('renders tasks when provided', () => {
    const tasks: Task[] = [
      { id: 1, title: 'Task 1', isCompleted: false, priority: 'p1' },
      { id: 2, title: 'Task 2', isCompleted: true, priority: 'p2' },
      { id: 3, title: 'Task 3', isCompleted: false, priority: 'p3' },
    ];

    const { container } = render(<TodoList {...defaultProps} tasks={tasks} />);

    expect(screen.getByTestId('todo-item-1')).toBeInTheDocument();
    expect(screen.getByTestId('todo-item-2')).toBeInTheDocument();
    expect(screen.getByTestId('todo-item-3')).toBeInTheDocument();

    expect(screen.getByText('Task 1')).toBeInTheDocument();
    expect(screen.getByText('Task 2')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Task 3')).toBeInTheDocument();
  });

  it('renders tasks in the order provided', () => {
    const tasks: Task[] = [
      { id: 3, title: 'Third task', isCompleted: false, priority: 'p1' },
      { id: 1, title: 'First task', isCompleted: false, priority: 'p2' },
      { id: 2, title: 'Second task', isCompleted: true, priority: 'p3' },
    ];

    render(<TodoList {...defaultProps} tasks={tasks} />);

    const todoItems = screen.getAllByTestId(/todo-item-/);
    expect(todoItems).toHaveLength(3);

    expect(todoItems[0]).toHaveAttribute('data-testid', 'todo-item-3');
    expect(todoItems[1]).toHaveAttribute('data-testid', 'todo-item-1');
    expect(todoItems[2]).toHaveAttribute('data-testid', 'todo-item-2');
  });

  it('updates when tasks prop changes', () => {
    const initialTasks: Task[] = [{ id: 1, title: 'Task 1', isCompleted: false, priority: 'p1' }];

    const { rerender } = render(<TodoList {...defaultProps} tasks={initialTasks} />);

    expect(screen.getByTestId('todo-item-1')).toBeInTheDocument();
    expect(screen.queryByText('No tasks yet!')).not.toBeInTheDocument();

    // Update to empty tasks
    rerender(<TodoList {...defaultProps} tasks={[]} />);

    expect(screen.queryByTestId('todo-item-1')).not.toBeInTheDocument();
    expect(screen.getByText('No tasks yet!')).toBeInTheDocument();

    // Update to new tasks
    const newTasks: Task[] = [{ id: 2, title: 'New Task', isCompleted: true, priority: 'p3' }];

    rerender(<TodoList {...defaultProps} tasks={newTasks} />);

    expect(screen.getByTestId('todo-item-2')).toBeInTheDocument();
    expect(screen.getByText('New Task')).toBeInTheDocument();
    expect(screen.queryByText('No tasks yet!')).not.toBeInTheDocument();
  });
});
