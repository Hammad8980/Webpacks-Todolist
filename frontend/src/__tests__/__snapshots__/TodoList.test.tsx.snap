// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TodoList renders tasks when provided 1`] = `
<div>
  <div
    class="space-y-3"
  >
    <div
      data-testid="todo-item-1"
    >
      Task 1
    </div>
    <div
      data-testid="todo-item-2"
    >
      Task 2
    </div>
    <div
      data-testid="todo-item-3"
    >
      Task 3
    </div>
  </div>
</div>
`;

exports[`TodoList shows empty state when no tasks are provided 1`] = `
<div>
  <div
    class="text-center py-12"
  >
    <div
      class="mb-4"
    >
      <svg
        class="w-16 h-16 text-gray-600 mx-auto"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
    <p
      class="text-gray-400 text-lg"
    >
      No tasks yet!
    </p>
    <p
      class="text-gray-500 text-sm mt-1"
    >
      Add your first task above to get started.
    </p>
  </div>
</div>
`;
